src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java

```diff
@@ -0,0 +1,300 @@
+     * @return ReadPair object holding the reads, or null if there are no more reads in the iterator
+     */
+    private ReadPair getNextReadPair(final PeekableIterator<SAMRecord> iterator) {
+
+        final ReadPair readPair = new ReadPair();
+        readPair.read1 = getNextUsableRead(iterator, false);
+        if (readPair.read1 == null) {
+            return null;
+        }
+
+        final SAMRecord peekedNextRead = getNextUsableRead(iterator, true);
+        if (peekedNextRead != null && peekedNextRead.getReadName().equals(readPair.read1.getReadName())) {
+            readPair.read2 = getNextUsableRead(iterator, false);
+        }
+
+        return readPair;
+    }
+
+    /**
+     * Return the next usable read in the iterator
+     *
+     * @param iterator the iterator to pull from
+     * @param justPeek if true, just peek the next read rather than pulling it
+     * @return the next read or null if none are left
+     */
+    private SAMRecord getNextUsableRead(final PeekableIterator<SAMRecord> iterator, final boolean justPeek) {
+
+        while (iterator.hasNext()) {
+            // trash the next read if it fails PF, is secondary, or is supplementary
+            final SAMRecord nextRead = iterator.peek();
+            if (nextRead.getReadFailsVendorQualityCheckFlag() || nextRead.isSecondaryOrSupplementary()) {
+                iterator.next();
+            }
+            // otherwise, return it
+            else {
+                return justPeek ? nextRead : iterator.next();
+            }
+        }
+
+        // no good reads left
+        return null;
+    }
+
+    private static class ReadPair {
+        public SAMRecord read1 = null;
+        public SAMRecord read2 = null;
+    }
+
+    /**
+     * Determine how many bases are excluded because of low mapping or base quality.
+     *
+     * @param read the read
+     * @return non-null object
+     */
+    private BaseExclusionHelper determineBaseExclusions(final SAMRecord read) {
+        final BaseExclusionHelper exclusions = new BaseExclusionHelper(read);
+
+        if (read.getMappingQuality() < MINIMUM_MAPPING_QUALITY) {
+            exclusions.basesExcludedByMapq = read.getReadLength();
+        } else {
+            exclusions.lowBQcount = getLowQualityOrSoftclipBaseCount(exclusions);
+        }
+
+        return exclusions;
+    }
+
+    private static class BaseExclusionHelper {
+        public SAMRecord read;
+        public int firstUnclippedBaseIndex;
+        public int firstTrailingClippedBaseIndex;
+        public int basesExcludedByMapq = 0;
+        public int lowBQcount = 0;
+
+        public BaseExclusionHelper(final SAMRecord read) {
+            this.read = read;
+
+            final List<CigarElement> cigarElements = read.getCigar().getCigarElements();
+            for (final CigarElement element : cigarElements) {
+                final CigarOperator op = element.getOperator();
+
+                if (op == CigarOperator.SOFT_CLIP) {
```

Comment:
This will not work when the read is hardclipped at the end. I think you need to switch on the op and break in case of a non-clip operator. also break after a softclip.

I'd add a test case or two 
1H1S1M -> 1
1S1M1S1H -> 1 // This currently fails, I think
1M1S -> 0 // this also currently fails, I think.

The lower part seems to have similar issues, in reverse.